File: Java/Merge_sort.java
class Sort 
{
    void merge(int arr[], int left, int middle, int right)
    {
        int low = middle - left + 1;                    
        int high = right - middle;                      
 
        int L[] = new int[low];                             
        int R[] = new int[high];

        int i = 0, j = 0;
 
        for (i = 0; i < low; i++)                             
        {
            L[i] = arr[left + i];
        }
        for (j = 0; j < high; j++)                            
        {
            R[j] = arr[middle + 1 + j];
        }
 
        int k = left;                                        
        i = 0;                                           
        j = 0;

        while (i < low && j < high)             
        {
            if (L[i] <= R[j]) 
            {
                arr[k] = L[i];
                i++;
            }
            else 
            {
                arr[k] = R[j];
                j++;
            }
            k++;
        }
        while (i < low)                         
        {
            arr[k] = L[i];
            i++;
            k++;
        }
        while (j < high)                          
        {
            arr[k] = R[j];
            j++;
            k++;
        }
    }

    void mergeSort(int arr[], int left, int right)    
    {
        int middle;
        if (left < right) {                         
            middle = (left + right) / 2;
            mergeSort(arr, left, middle);               
            mergeSort(arr, middle + 1, right);           
            merge(arr, left, middle, right);          
        }
    }
 
    void display(int arr[])      
    {  
        for (int i=0; i<arr.length; ++i) 
        {
            System.out.print(arr[i]+" ");
        } 
    } 

    public static void main(String args[])
    {
        int arr[] = { 2, 8, 11, 15, 43, 12 };
        Sort ob = new Sort();
        ob.mergeSort(arr, 0, arr.length - 1);
        ob.display(arr);
    }
}

File: C/merge_sort.c
#include<stdio.h>

// MERGE SORT

void merge(int arr[], int low, int mid, int high) {
	int n1 = mid - low + 1;
	int n2 = high - mid;

	int L[n1], R[n2], i, j, k;

	for (i = 0; i < n1; i++)
		L[i] = arr[low + i];

	for (j = 0; j < n2; j++)
		R[j] = arr[mid + 1 + j];

	i = j = 0;
	k = low;

	while (i < n1 && j < n2) {
		if (L[i] <= R[j]) {
			arr[k] = L[i];
			i++;
		} else {
			arr[k] = R[j];
			j++;
		}
		k++;
	}

	while (i < n1) {
		arr[k] = L[i];
		k++;
		i++;
	}
	while (j < n2) {
		arr[k] = R[j];
		k++;
		j++;
	}
}

void merge_sort(int arr[], int low, int high) {
	if (low < high) {
		int mid = low + (high - low) / 2;
		merge_sort(arr, low, mid);
		merge_sort(arr, mid + 1, high);
		merge(arr, low, mid, high);
	}
}

int main()
{
	int arr[] = { 80, 90, 100, 40, 50, 30, 20, 70, 60, 10}, n = 10;

	merge_sort(arr, n);

	print_arr(arr, n);

	return 0;
}